Refactor retry download

andre2007 opened the pull request:
The logic for download retry was duplicated. This pull requests moves the retry logic into the internal
utils module. In addition the old logic had the flaw, in case of CurlException no retry was done.

WebFreak001 commented on source/dub/packagesuppliers/maven.d:
```diff
@@ -92,27 +88,16 @@ class MavenRegistryPackageSupplier : PackageSupplier {
+				return Json(null);
 			}
+			else throw e;
 		}
```
this now behaves differently. Before it would have just tried to parse an empty xml after 3 fails and put that in the cache, but now it throws after 3 fails. Though this seems like it was a bug before so I think this is a good change here

WebFreak001 commented on source/dub/packagesuppliers/registry.d:
```diff
@@ -92,28 +87,18 @@ class RegistryPackageSupplier : PackageSupplier {
+			else throw e;
 		}
+
 		Json json = parseJsonString(jsonData, url.toString());
```
same as above

WebFreak001 commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,83 @@ ubyte[] download(URL url, uint timeout = 8)
 }
 
+/**
+	Downloads a file from the specified URL with retry logic.
```
I think the documentation could be improved here. Something like
```
Downloads a file from the specified URL with up to 3 trys on failure
Throws: `Exception` if the download failed or `HTTPStatusException` on "unrecoverable failures" such as 404 not found. Otherwise might throw anything else that `download` throws.
See_Also: download
```

andre2007 replied:
done

andre2007 commented:
@WebFreak001 thanks for your feedback, I added the description.

I also corrected the 2 fetchPackage functions. In case  a HTTPStatusException with status code != 404 was caught, no new Exception was thrown.

WebFreak001 commented on source/dub/packagesuppliers/registry.d:
```diff
@@ -57,19 +57,16 @@ class RegistryPackageSupplier : PackageSupplier {
+		}
+		catch(HTTPStatusException e) {
+			if (e.status == 404) throw e;
+			else logDebug("Failed to download package %s from %s", packageId, url); 
```
this code can't trigger, if a HTTPStatusException is ever thrown it's only a 404 one (because you catch the rest). Only at the end you throw a normal Exception. Maybe rethrow the last HTTPStatusException on failure after 3 retries instead?

andre2007 replied:
HTTPStatusException is now rethrown
The throw Exception on last line of retryDownload I kept in, because of the CurlException

wilzbach commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,92 @@ ubyte[] download(URL url, uint timeout = 8)
+					else continue;
+				}
+			}
+			catch(CurlException e) {
```
Why not putting this only in a `version` block?

wilzbach commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,92 @@ ubyte[] download(URL url, uint timeout = 8)
+					else continue;
+				}
+			}
+			catch(CurlException e) {
```
Why not putting this only in a `version` block?

andre2007 replied:
I am not sure whether I understood your question correctly.
Did you asked why I didn't put the the catch(CurlException) only
In a version scope?
I tried this but got syntax errors. It seems it isn't allowed.

wilzbach commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,92 @@ ubyte[] download(URL url, uint timeout = 8)
+		}
+	}
+	throw new Exception("Failed to download %s".format(url));
+}
```
One empty line between functions

andre2007 replied:
done

wilzbach commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,92 @@ ubyte[] download(URL url, uint timeout = 8)
+**/
+void retryDownload(URL url, NativePath filename)
+{
+	foreach(i; 0..3) {
```
Probably a good idea to make n configurable if we already generalize this function?

andre2007 replied:
done

andre2007 commented on source/dub/internal/utils.d:
```diff
@@ -325,6 +325,93 @@ ubyte[] download(URL url, uint timeout = 8)
+	Downloads a file from the specified URL with retry logic.
+
+	Downloads a file from the specified URL with up to n tries on failure
+	Throws: `Exception` if the download failed or `HTTPStatusException` on "unrecoverable failures" such as 404 not found.
```
done

andre2007 commented:
Ping :)
